File: src/main/java/prog/CompileRefs.java
package prog;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public class CompileRefs {
    private static final Path SRC_ROOT = Paths.get("src", "main", "java").toAbsolutePath().normalize();

    /** Mirrors the file structure of .../obj/ into .../ref/, and automatically writes (or delete) the reference classes. */
    static void main(String[] args) {
        new CompileRefs().run();
    }

    private void run() {
        String objPath = "src/main/java/prog/obj/";
        String refPath = "src/main/java/prog/ref/";

        File objDir = new File(objPath);
        File refDir = new File(refPath);

        if (!objDir.exists()) { boolean _ = objDir.mkdirs(); }
        if (!refDir.exists()) { boolean _ = refDir.mkdirs(); }

        explore(objPath);
        cleanup(refPath);
    }

    /**
     * Recursively explores the given directory, creating reference classes and mirroring new directories.
     * @param currentObjPath the current object path
     */
    private void explore(String currentObjPath) {
        String currentRefPath = currentObjPath.replaceFirst("obj", "ref");
        File currentObjDir = new File(currentObjPath);
        File currentRefDir = new File(currentRefPath);

        if (!currentRefDir.exists()) { boolean _ = currentRefDir.mkdirs(); }

        File[] files = currentObjDir.listFiles();
        if (files == null) return;

        for (File file: files) {
            if (file.isDirectory()) explore(file.getPath());
            else writeRef(file);
        }
    }

    /**
     * Recursively explore the given directory, deleting any reference classes and directories the no longer have a corresponding object class or directory.
     * @param currentRefPath the current ref path
     */
    private void cleanup(String currentRefPath) {
        String currentObjPath = currentRefPath.replaceFirst("ref", "obj");
        File currentObjDir = new File(currentObjPath);
        File currentRefDir = new File(currentRefPath);

        if (!currentObjDir.exists()) { boolean _ = currentRefDir.delete(); }

        File[] files = currentRefDir.listFiles();
        if (files == null) return;

        for (File file: files) {
            if (file.isDirectory()) cleanup(file.getPath());
            else {
                String objFilePath = file.getPath().replaceFirst("ref", "obj").replaceFirst("Ref\\.java$", ".java");
                if (!new File(objFilePath).exists()) { boolean _ = file.delete(); }
            }
        }
    }

    private void writeRef(File file){
        Path objPath = file.toPath().toAbsolutePath().normalize();
        Path objRoot = Paths.get("src", "main", "java", "prog", "obj").toAbsolutePath().normalize();
        Path refRoot = Paths.get("src", "main", "java", "prog", "ref").toAbsolutePath().normalize();

        Path relObjPath = objRoot.relativize(objPath);
        String objClassName = relObjPath.getFileName().toString().replaceFirst("\\.java$", "");
        String refClassName = objClassName + "Ref";

        Path refFilePath = refRoot.resolve(relObjPath).resolveSibling(refClassName + ".java");
        String pkgName = packageFromPath(refFilePath);

        Path objFilePath = objRoot.resolve(relObjPath).resolveSibling(objClassName + ".java");
        String importName = packageFromPath(objFilePath) + "." + objClassName;

        String content =
                "package " + pkgName + ";\n" +
                "\n" +
                "import language.Ref;\n" +
                "\n" +
                "import " + importName + ";\n" +
                "\n" +
                "/** Represents a reference to a " + objClassName + " value */\n" +
                "public class " + refClassName + " extends Ref<" + objClassName + "> {\n" +
                "    private " + objClassName + " field;\n" +
                "\n" +
                "    /** Sets the referenced value. */\n" +
                "    @Override\n" +
                "    public void set(" + objClassName + " value) { field = value; }\n" +
                "\n" +
                "    /** @return the referenced value. */\n" +
                "    @Override\n" +
                "    public " + objClassName + " get() { return field; }\n" +
                "\n" +
                "    /** @return a reference to the given " + objClassName + ". */\n" +
                "    public static " + refClassName + " of(" + objClassName + " value) {\n" +
                "        " + refClassName + " ref = new " + refClassName + "();\n" +
                "        ref.field = value;\n" +
                "        return ref;\n" +
                "    }\n" +
                "\n" +
                "/** @return a reference to a copy of the referenced value. */\n" +
                "    @Override\n" +
                "    public " + refClassName + " copy() { return " + refClassName + ".of(field.copy()); }\n" +
                "}\n";

        try {
            Files.writeString(refFilePath, content, StandardCharsets.UTF_8);
        } catch (IOException e) {
            System.err.println("Failed to write reference class for " + objClassName + ": " + e.getMessage());
        }
    }

    private static String packageFromPath(Path filePath) {
        Path abs = filePath.toAbsolutePath().normalize();
        Path relToSrc = SRC_ROOT.relativize(abs.getParent());
        return relToSrc.toString().replace('\\', '.').replace('/', '.');
    }
}

File: src/main/java/language/CompileRefs.py
import os
import sys
from pathlib import Path

class CompileRefs:
    def __init__(self):
        self.src_root = Path("/").resolve()

    def run(self):
        obj_path = "obj/"
        ref_path = "ref/"

        obj_dir = Path(obj_path)
        ref_dir = Path(ref_path)

        obj_dir.mkdir(parents=True, exist_ok=True)
        ref_dir.mkdir(parents=True, exist_ok=True)

        self.explore(obj_path)
        self.cleanup(ref_path)

    def explore(self, current_obj_path):
        """Recursively explores the given directory, creating reference classes and mirroring new directories."""
        current_ref_path = current_obj_path.replace("obj", "ref")
        current_obj_dir = Path(current_obj_path)
        current_ref_dir = Path(current_ref_path)

        current_ref_dir.mkdir(parents=True, exist_ok=True)

        try:
            files = list(current_obj_dir.iterdir())
        except (FileNotFoundError, NotADirectoryError):
            return

        for file in files:
            if file.is_dir():
                self.explore(str(file))
            else:
                self.write_ref(file)

    def cleanup(self, current_ref_path):
        """Recursively explore the given directory, deleting any reference classes and directories that no longer have a corresponding object class or directory."""
        current_obj_path = current_ref_path.replace("ref", "obj")
        current_obj_dir = Path(current_obj_path)
        current_ref_dir = Path(current_ref_path)

        if not current_obj_dir.exists():
            try:
                current_ref_dir.rmdir()
            except OSError:
                pass
            return

        try:
            files = list(current_ref_dir.iterdir())
        except (FileNotFoundError, NotADirectoryError):
            return

        for file in files:
            if file.is_dir():
                self.cleanup(str(file))
            else:
                obj_file_path = str(file).replace("ref", "obj").replace("Ref.java", ".java")
                if not Path(obj_file_path).exists():
                    try:
                        file.unlink()
                    except OSError:
                        pass

    def write_ref(self, file):
        file_path = file.resolve()
        obj_root = Path("obj/").resolve()
        ref_root = Path("ref/").resolve()

        rel_obj_path = file_path.relative_to(obj_root)
        obj_class_name = rel_obj_path.stem
        ref_class_name = obj_class_name + "Ref"

        ref_file_path = ref_root / rel_obj_path.parent / (ref_class_name + ".java")
        pkg_name = self.package_from_path(ref_file_path)

        obj_file_path =  obj_root / rel_obj_path
        import_name = self.package_from_path(obj_file_path) + "." + obj_class_name

        content = f"""package {pkg_name};

import language.Ref;

import {import_name};

/** Represents a reference to a {obj_class_name} value */
public class {ref_class_name} extends Ref<{obj_class_name}> {{
    private {obj_class_name} field;

    /** Sets the referenced value. */
    @Override
    public void set({obj_class_name} value) {{ field = value; }}

    /** @return the referenced value. */
    @Override
    public {obj_class_name} get() {{ return field; }}

    /** @return a reference to the given {obj_class_name}. */
    public static {ref_class_name} of({obj_class_name} value) {{
        {ref_class_name} ref = new {ref_class_name}();
        ref.field = value;
        return ref;
    }}

/** @return a reference to a copy of the referenced value. */
    @Override
    public {ref_class_name} copy() {{ 
        if (field != null) return {ref_class_name}.of(field.copy());
        return new {ref_class_name}();
    }}
}}
"""

        try:
            ref_file_path.parent.mkdir(parents=True, exist_ok=True)
            ref_file_path.write_text(content, encoding='utf-8')
        except IOError as e:
            print(f"Failed to write reference class for {obj_class_name}: {e}", file=sys.stderr)

    @staticmethod
    def package_from_path(file_path):
        abs_path = file_path.resolve()
        src_root = Path("").resolve()
        rel_to_src = abs_path.parent.relative_to(src_root)
        return "language." + str(rel_to_src).replace('\\', '.').replace('/', '.')


def main():
    CompileRefs().run()


if __name__ == "__main__":
    main()